File: Python/0008.py
class Solution:
    def myAtoi(self, s: str) -> int:
        # Define 32-bit integer limits
        INT_MIN = -2**31
        INT_MAX = 2**31 - 1
        
        # Initialize variables
        result = 0
        sign = 1
        i = 0
        
        # Skip leading whitespace
        while i < len(s) and s[i] == ' ':
            i += 1
        
        # Handle sign
        if i < len(s) and (s[i] == '-' or s[i] == '+'):
            if s[i] == '-':
                sign = -1
            i += 1
        
        # Convert digits
        while i < len(s) and s[i].isdigit():
            digit = int(s[i])
            # Check for overflow before adding new digit
            if result > (INT_MAX - digit) // 10:
                return INT_MAX if sign == 1 else INT_MIN
            result = result * 10 + digit
            i += 1
        
        # Apply sign
        result *= sign
        
        # Ensure result is within the 32-bit integer range
        if result < INT_MIN:
            return INT_MIN
        elif result > INT_MAX:
            return INT_MAX
        
        return result

File: C/0008.c
#include <limits.h> // For INT_MAX and INT_MIN
#include <ctype.h>  // For isdigit() and isspace()

int myAtoi(char* s) {
    int i = 0;
    int sign = 1;
    long result = 0;

    // Step 1: Ignore leading whitespaces
    while (isspace(s[i])) {
        i++;
    }

    // Step 2: Check the sign
    if (s[i] == '-' || s[i] == '+') {
        sign = (s[i] == '-') ? -1 : 1;
        i++;
    }

    // Step 3: Convert digits to integer and handle overflow
    while (isdigit(s[i])) {
        result = result * 10 + (s[i] - '0');

        // Step 4: Handle overflow
        if (result * sign >= INT_MAX) {
            return INT_MAX;
        } else if (result * sign <= INT_MIN) {
            return INT_MIN;
        }

        i++;
    }

    // Return the final result with the correct sign
    return (int)(result * sign);
}
